### PR TITLE
✨Amp Form - Async Input / amp-recaptcha-input Support

### LINKED ISSUE #18829: <amp-recaptcha-input> Tracking Issue for Remaning Steps For Experimental
relates to #2273

* [x] Origin Domain Verification
* [x] Install Recaptcha Service on Document Level, Use Viewer Service's cached isProxyOrigin.
* [x] Check for same sitekey (throw user error)
* [ ] AMP Async Input

We also should consider letting recaptcha know, as well as the AMP validation team. And then writing docs / examples

### PR DESCRIPTION
**Pull request by torch2424**
relates to #2273
closes #18829

🎉  Final PR for `amp-recaptcha-input` before being code complete 🎉 

**Feature Changes:**

* Adds support for Async Input ( See design doc in #2273 ) inside of AMP Form. 
* Adds the `data-name` attribute for the Async Input interface for AMP From to use.

**TODO**

* ~~This still needs Unit tests. Currently this breaks some tests on AMP Form. But Wanted to decide on an implementation before fixing / adding new tests.~~

* ~~Error Handling when a `getValue()` call rejects. Should cancel the submission entirely, and put the form in an error state.~~

* Documentation for amp-recaptcha-input (In another PR)

# Example

Goes through Firefox and Safari.

recaptchaampform

Error:

screen shot 2018-11-20 at 2 13 28 pm

### CODE REVIEW

File: extensions/amp-form/0.1/amp-form.js [hunk 1]
```diff
@@ -525,18 +568,15 @@ export class AmpForm {
...
    * @private
    */
-  handleXhrSubmit_(varSubsFields, trust) {
-    this.setState_(FormState.SUBMITTING);
-    const varSubPromise = this.doVarSubs_(varSubsFields);
+  handleXhrSubmit_(trust) {
     let p;
     if (this.ssrTemplateHelper_.isSupported()) {
-      p = varSubPromise.then(() => this.handleSsrTemplate_(trust));
+      p = Promise.resolve().then(() => this.handleSsrTemplate_(trust));
```
Comment (cvializ) [inline extensions/amp-form/0.1/amp-form.js]: nit:
```js
// Do this...
const p = functionThatReturnsAPromise();

// ...instead of this
const p = Promise.resolve().then(() => functionThatReturnsAPromise())

```
You can save bytes and make the code cleaner by using the `Promise` from the function and omitting the `Promise.resolve()`.

File: extensions/amp-recaptcha-input/0.1/test/test-amp-recaptcha-input.js [hunk 2]
```diff
@@ -32,10 +32,14 @@ describes.realWin('amp-recaptcha-input', {
     toggleExperiment(win, 'amp-recaptcha-input', true);
   });
 
-  function getRecaptchaInput(skipSiteKey, skipAction) {
+  function getRecaptchaInput(skipName, skipSiteKey, skipAction) {
```
Comment (cvializ) [inline extensions/amp-recaptcha-input/0.1/test/test-amp-recaptcha-input.js]: nit: flag booleans in test methods can make the tests difficult to grok sometimes. Consider using a config object with these as properties instead. That way in the tests you can label the keys at the call sites

```js
// This can be easier to understand
getRecaptchaInput({skipAction: true});

// This can be harder to understand
getRecaptchaInput(undefined, undefined, true);

// This is better but not 💯
getRecaptchaInput(/*opt_skipName*/ undefined, /*opt_skipSiteKey*/ undefined, /*opt_skipAction*/ true);
```

File: src/async-input.js [hunk 3]
```diff
@@ -39,6 +39,27 @@ export class AsyncInput {
...
+ */
+export const AsyncInputAttributes = {
+  /**
+   * data-name
+   *
+   * Required attribute that must be asserted by every async-input
+   * Element. This is used by AMP form to add the key
+   * for the form submission request
+  */
+  'NAME': 'data-name',
```
Comment (cvializ) [inline src/async-input.js]: design nit: data attributes are typically used when creating new attributes for existing DOM elements, like `input` or `p`. This prevents new standards and specs from the W3C from being added to existing DOM elements like `input` and making your markup behave differently over time.

Since you're making a custom element `amp-recaptcha-input` that the W3C will never add new attributes to, the `data-*` prefix doesn't add the utility that it adds for existing web standard elements.

File: extensions/amp-form/0.1/amp-form.js [hunk 4]
```diff
@@ -465,19 +462,23 @@ export class AmpForm {
...
 
   /**
    * Helper method that actual handles the different cases (post, get, xhr...).
    * @param {ActionTrust} trust
+   * @param {?Event} event
    * @private
    */
-  submit_(trust) {
+  submit_(trust, event) {
+    // TODO(torch2424) remove this todo. Used to navigate code :p
```
Comment (cvializ) [inline extensions/amp-form/0.1/amp-form.js]: haha I do the same thing : P

File: extensions/amp-form/0.1/amp-form.js [hunk 5]
```diff
@@ -488,14 +489,56 @@ export class AmpForm {
...
+      if (asyncInputs.length <= 0) {
+
+        for (let i = 0; i < varSubsFields.length; i++) {
+          this.urlReplacement_.expandInputValueSync(varSubsFields[i]);
+        }
+
+        this.handleNonXhrGet_();
+        return;
+      } else if (event) {
+        event.stopImmediatePropagation();
```
Comment (cvializ) [inline extensions/amp-form/0.1/amp-form.js]: Is `stopImmediatePropagation` really needed here? It can become a footgun if other code is listening for the `submit` event, like the `form-submit-service`. The other place in this file that calls `stopImmediatePropagation` has a huge comment explaining its significance in the design.

File: extensions/amp-form/0.1/amp-form.js [hunk 6]
```diff
@@ -488,14 +489,56 @@ export class AmpForm {
...
+      // If we have no async inputs, we can just submit synchronously
+      if (asyncInputs.length <= 0) {
+
+        for (let i = 0; i < varSubsFields.length; i++) {
+          this.urlReplacement_.expandInputValueSync(varSubsFields[i]);
+        }
+
+        this.handleNonXhrGet_();
+        return;
+      } else if (event) {
```
Comment (cvializ) [inline extensions/amp-form/0.1/amp-form.js]: style nit: It's cleaner to omit the `else` if the `if` before it calls return.

File: extensions/amp-form/0.1/amp-form.js [hunk 7]
```diff
@@ -488,14 +489,56 @@ export class AmpForm {
...
+    // Do any assertions we may need to do
+    // For NonXhrGET
+    // took out all the code in handleNonXhrGet_
+    // because alot was assertions we want to fail on before getting var subs
+    if (!this.xhrAction_ && this.method_ == 'GET') {
+      this.assertSsrTemplate_(false, 'Non-XHR GETs not supported.');
+      this.assertNoSensitiveFields_();
+
+      // If we have no async inputs, we can just submit synchronously
+      if (asyncInputs.length <= 0) {
```
Comment (cvializ) [inline extensions/amp-form/0.1/amp-form.js]: style nit: Will the length ever be less than `0`? You can use `==` to say exactly what you mean

File: extensions/amp-form/0.1/amp-form.js [hunk 8]
```diff
@@ -409,10 +409,7 @@ export class AmpForm {
       return;
     }
     // `submit` has the same trust level as the AMP Action that caused it.
-    this.submit_(invocation.trust);
-    if (this.method_ == 'GET' && !this.xhrAction_) {
-      this.form_.submit();
```
Comment (cvializ) [inline extensions/amp-form/0.1/amp-form.js]: Think very carefully about changing the order of significant control flows like this. Make sure to verify that the `method="GET" action="https://example.com/form` success and error cases still work as expected after this change to the submit action.

File: extensions/amp-form/0.1/amp-form.js (hunk 1 above)
Comment (torch2424) [reply extensions/amp-form/0.1/amp-form.js]: Oops, yeah totally, thanks! 😄

File: extensions/amp-recaptcha-input/0.1/test/test-amp-recaptcha-input.js (hunk 2 above)
Comment (torch2424) [reply extensions/amp-recaptcha-input/0.1/test/test-amp-recaptcha-input.js]: Yeah it got kinda out of hand. Thanks! Will do this.

File: src/async-input.js (hunk 3 above)
Comment (torch2424) [reply src/async-input.js]: Ahh so since I am passing things through a 3p frame, it seems that all data you want to be passed through the 3p frame, we have been prepending with `data-*`. See https://www.ampproject.org/docs/reference/components/amp-gist . and we already have `data-sitekey` and `data-action`

Originally I thought it would be better to be consistent. But now that I realize that layout is also not `data-`, I can agree with this 😄 Thanks!

File: extensions/amp-form/0.1/amp-form.js (hunk 5 above)
Comment (torch2424) [reply extensions/amp-form/0.1/amp-form.js]: Oops. Thank you for the catch!

File: extensions/amp-form/0.1/amp-form.js (hunk 8 above)
Comment (torch2424) [reply extensions/amp-form/0.1/amp-form.js]: Yeah that is one thing I definitely tried to be careful of. I have an example of all the flow cases this affects. I tested it manually and all of it seems fine. Havent tested an error state yet though, simply because I don't know a good Endpoint to hit to test this. Do we have one that just returns 500s or something?

**Comment by torch2424** [comment]
@cvializ Made your requested changes 😄

File: extensions/amp-form/0.1/amp-form.js (hunk 8 above)
Comment (cvializ) [reply extensions/amp-form/0.1/amp-form.js]: That should work. There are a couple of endpoints in `app.js` that serve `statusCode`s like `400` or `500`

File: extensions/amp-form/0.1/amp-form.js [hunk 9]
```diff
@@ -652,6 +694,33 @@ export class AmpForm {
...
+  /**
+   * Call getValue() on Async Input elements, and
+   * Create hidden inputs containing their returned values
+   * @param {!Element} asyncInput
+   * @return {!Promise}
+   * @private
+   */
+  getValueForAsyncInput_(asyncInput) {
+    return asyncInput.getImpl().then(implementation => {
+      return implementation.getValue().then(value => {
```
Comment (cvializ) [inline extensions/amp-form/0.1/amp-form.js]: nit:`Promise` chains should almost always be simplified to be one-level deep.

```js
return asyncInput.getImpl()
    .then(implementation => implementation.getValue())
    .then(value => {
      /* ... */
    });
```
 This avoids the Promise Pyramid of Doom. 

> ["You can easily smell a promise pyramid whenever you see then() blocks going inwards in an existing then() block."](https://web.archive.org/web/20180301090527/http://blog.ministryofprogramming.com/prevent-pyramids-with-flat-promise-chains/)

File: extensions/amp-form/0.1/amp-form.js [hunk 10]
```diff
@@ -488,14 +493,57 @@ export class AmpForm {
...
+        this.form_.querySelectorAll(
+            `.${escapeCssSelectorIdent(AsyncInputClasses['ASYNC_INPUT'])}`
+        )
+    );
+
+    // Do any assertions we may need to do
+    // For NonXhrGET
+    // took out all the code in handleNonXhrGet_
+    // because alot was assertions we want to fail on before getting var subs
+    if (!this.xhrAction_ && this.method_ == 'GET') {
```
Comment (cvializ) [inline extensions/amp-form/0.1/amp-form.js]: I don't understand why this change was made to `handleNonXhrGet`. Is it an optimization to avoid waiting for the the `Promise.all` call in case there are no `asyncInput`s? If that is the case, this `if` block is redundant because the `Promise.all` call should be a noop if it is passed an empty array.

Why do we want the assertions to fail before getting var subs?

File: extensions/amp-recaptcha-input/0.1/test/test-amp-recaptcha-input.js [hunk 11]
```diff
@@ -32,19 +32,35 @@ describes.realWin('amp-recaptcha-input', {
...
 
-  function getRecaptchaInput(skipSiteKey, skipAction) {
+  function getRecaptchaInput(config) {
     const ampRecaptchaInput = doc.createElement('amp-recaptcha-input');
     ampRecaptchaInput.setAttribute('layout',
         'nodisplay');
-    if (!skipSiteKey) {
+
+    if (config) {
+      if (!config.skipName) {
```
Comment (cvializ) [inline extensions/amp-recaptcha-input/0.1/test/test-amp-recaptcha-input.js]: nit: [Flat is better than nested](https://en.wikibooks.org/wiki/Computer_Programming/Coding_Style/Minimize_nesting). Avoid nesting if blocks, and nesting blocks generally. It tends to make code harder to understand.



```js
function getRecaptchaInput(config = {}) {
  /* ... */

  if (!config.skipName) {
    ampRecaptchaInput.setAttribute('name', 'recaptcha-test');
  }

  /* ... */
}
```

File: src/async-input.js [hunk 12]
```diff
@@ -39,6 +39,27 @@ export class AsyncInput {
...
 }
 
+/**
+ * Attributes
+ *
+ * Components implementing the AsyncInput,
+ * are expected to assert these attributes
+ * at the appropriate time.
+ *
+ * @constant {!Object<string, string>}
```
Comment (cvializ) [inline src/async-input.js]: This is an `@enum {string}` and you can reference it with dot-notation. Closure Compiler should mangle the property name, but not the value.

File: extensions/amp-form/0.1/amp-form.js (hunk 10 above)
Comment (torch2424) [reply extensions/amp-form/0.1/amp-form.js]: Yes, you mostly right 😄 

I moved this out in order to to be able to check all of the assertions for NonXHRGet. We want them to "fail" (bad choice of words on my part), so we can determine if we can submit synchronously vs async. 

So, while it is also an optimization, I mostly did this so that if we don't have any async inputs, then we know we can submit sync. Even though the Promise.all will noop, I am pretty sure that it will still be popped onto the task queue, thus making it async, and it won't synchronously run the `.then` block. Please correct me if I am wrong 😄

File: extensions/amp-recaptcha-input/0.1/test/test-amp-recaptcha-input.js (hunk 11 above)
Comment (torch2424) [reply extensions/amp-recaptcha-input/0.1/test/test-amp-recaptcha-input.js]: Didn't know our compiler would handle the default param assignment thingy (don't know the technical term off of the top of my head haha!), will definitely use that, thanks! 😄

File: src/async-input.js (hunk 12 above)
Comment (torch2424) [reply src/async-input.js]: cc @zhouyx Didn't know we could do this! 😄 

Thanks @cvializ !

**Comment by torch2424** [comment]
@cvializ Made requested changes / Replied to comments  😄

File: extensions/amp-recaptcha-input/0.1/test/test-amp-recaptcha-input.js (hunk 11 above)
Comment (cvializ) [reply extensions/amp-recaptcha-input/0.1/test/test-amp-recaptcha-input.js]: np! you can do the same thing with `config = config || {};`

File: src/async-input.js (hunk 12 above)
Comment (cvializ) [reply src/async-input.js]: yes! `@enums` are great since I think they compile down to just the `string` value, which lets Closure Compiler get rid of the object reference.

File: extensions/amp-form/0.1/amp-form.js (hunk 10 above)
Comment (cvializ) [reply extensions/amp-form/0.1/amp-form.js]: Cool, yeah you're right about the task queue. If you keep this structure, does it still make sense for the non-XHR get to be in the block below?

https://github.com/ampproject/amphtml/blob/08720df6fabea59cc2a1f032b8fe46bb3374130d/extensions/amp-form/0.1/amp-form.js#L543-L545

**Comment by cvializ** [comment]
Thanks for addressing my comments!! I'll give LGTM when Travis passes

File: extensions/amp-form/0.1/amp-form.js (hunk 10 above)
Comment (torch2424) [reply extensions/amp-form/0.1/amp-form.js]: Rad!

And yes 😄 The blow above that calls `this.handleNonXhrGet_();` will handle the sync case. This will call `this.handleNonXhrGet_();` in the async case once we are done making all of our hidden inputs and things 😄

**Comment by torch2424** [comment]
@cvializ Awesome! Thanks for looking over this! Glad we are in a LGTM-able state So here are the remaining things we need to do in this PR then:

* Error Handling

* Fix Current Tests

* Write new Unit Tests

I'll go ahead and do the error handling, ask you to look over that, and then I'll start fixing / writing tests and getting travis passing 😄

File: extensions/amp-form/0.1/amp-form.js [hunk 13]
```diff
@@ -488,14 +493,59 @@ export class AmpForm {
...
+
+    Promise.all(presubmitPromises).then(() => {
+      if (this.xhrAction_) {
+        this.handleXhrSubmit_(trust);
+      } else if (this.method_ == 'POST') {
+        this.handleNonXhrPost_();
+      } else if (this.method_ == 'GET') {
+        this.handleNonXhrGet_();
+      }
+    }).catch(error => {
```
Comment (torch2424) [inline extensions/amp-form/0.1/amp-form.js]: Added error handling here.

File: examples/recaptcha.amp.html [hunk 14]
```diff
@@ -37,23 +90,164 @@ <h1>Amp Recaptcha Examples</h1>
...
+        </div>
+
+        <div submit-success>
+          <template type="amp-mustache">
+            <h1>You searched for: {{term}}</h1>
+            <p>Recaptcha token: {{recaptcha_example}}</p>
+          </template>
+        </div>
+
+        <div submit-error>
```
Comment (torch2424) [inline examples/recaptcha.amp.html]: And here's an example

**Comment by torch2424** [comment]
@cvializ Added error handling support 😄 

I went ahead and left some comments on some lines to handle the error handling.

And here is a screenshot I made by throwing an error in getValue.

screen shot 2018-11-19 at 2 45 37 pm

Let me know if this looks good and I can start on the tests and things 😄

File: extensions/amp-form/0.1/amp-form.js (hunk 13 above)
Comment (cvializ) [reply extensions/amp-form/0.1/amp-form.js]: This should be unnecessary: the error is already handled by `handleXhrSubmitFailure_`. The non-xhr cases don't need to be handled by AMP since the browser redirects in the non-xhr cases.

Comment (cvializ) [reply extensions/amp-form/0.1/amp-form.js]: Consider splitting the `then`/`catch` clauses here into two functions, `handlePresubmitSuccess_` and `handlePresubmitFailure_` and call them in the first and second arguments to the `then`. The difference between `promise.then(successFn).catch(errorFn)` and `promise.then(successFn, errorFn)` and is that the `catch` block in the first option will receive errors from the `then` or from the root promise, where the second option will only receive errors from the promise.

```js
Promise.all(presubmitPromises).then(
    () => this.handlePresubmitSuccess_(trust),
    error => this.handlePresubmitError_(error);
```

`handleXhrSubmit_` already handles its own errors, so something like this makes it clearer that the error handling code is only designed to respond to the presubmit promises, and not any of the methods in the success block

File: extensions/amp-form/0.1/amp-form.js [hunk 15]
```diff
@@ -652,6 +696,33 @@ export class AmpForm {
...
+  /**
+   * Call getValue() on Async Input elements, and
+   * Create hidden inputs containing their returned values
+   * @param {!Element} asyncInput
+   * @return {!Promise}
+   * @private
+   */
+  getValueForAsyncInput_(asyncInput) {
+    return asyncInput.getImpl()
+        .then(implementation => implementation.getValue())
```
[the remark this answers is not in the document]
Comment (cvializ) [reply extensions/amp-form/0.1/amp-form.js]: Good point @zhouyx it could make sense to add a timeout for a better submit error UX. The async variable substitution has a timeout of `100`ms, so a similar wait for the async inputs would make sense.

[the remark this answers is not in the document]
Comment (torch2424) [reply extensions/amp-form/0.1/amp-form.js]: Just to clarify, are we suggesting using `this.waitOnPromisesOrTimeout_` ? And the timeout is just for if the request fails silently and doesn't resolve/reject ?

What would be a good timeout value? I noticed recaptcha takes a few seconds to return a token. And any other future async input elements (google pay), I am sure could take even longer. I think 100 would be too short? But I do agree some sort of timeout would be nice.

**Comment by cvializ** [comment]
Feel free to start on tests

File: extensions/amp-form/0.1/amp-form.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (cvializ) [reply extensions/amp-form/0.1/amp-form.js]: I believe so, @zhouyx ? 

A good value depends on how long the user should expect the response to take after clicking submit. Does the recaptcha request start at page load? If so, even if the recaptcha api response takes some time, the user probably won't submit the form until a number of seconds after the page loads. Just some sane upper-bound value that won't abandon ship too often works. 1s after submit is started? 3s? 5s? 10s?

[the remark this answers is not in the document]
Comment (zhouyx) [reply extensions/amp-form/0.1/amp-form.js]: @cvializ recaptcha create the iframe and install service at `layoutCallback`. However the token is generating real time when user click summit, and that could take longer than 100ms. There's the external js and network involved. @torch2424 has better idea on the typical latency. 

I think @torch2424 made a good point, that different services have different typical latency. Having the same timeout value doesn't make much sense for other services we introduce later on. Instead I believe we should require every service define their own timeout value with `getValue()`. Then we can also introduce a pretty long timeout (10s?) here for safety.

[the remark this answers is not in the document]
Comment (cvializ) [reply extensions/amp-form/0.1/amp-form.js]: That makes sense. I agree, a timeout specific to each AsyncInput and a long safety timeout sound like good ideas.

File: extensions/amp-form/0.1/amp-form.js (hunk 13 above)
Comment (torch2424) [reply extensions/amp-form/0.1/amp-form.js]: Sounds great, thanks for the detailed explanation! 😄

File: extensions/amp-form/0.1/amp-form.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (torch2424) [reply extensions/amp-form/0.1/amp-form.js]: Awesome. I agree, lets do an upper bound of 10 seconds, and encourage other Async inputs to make a shorter timeout if appropriate? @zhouyx @cvializ 

Personally, I'd want to just use the upper bound of 10s for recaptcha. Simply because it could take even longer on 2g and NBU users and things. Though, usually in my manual testing, it takes about 1-2 seconds.

**Comment by torch2424** [comment]
@cvializ 

I'm having a hard time getting the tests to handle the new asynchrony of the submit event. Since we now wait for all promises to resolve before calling the corresponding `handleXhrSubmit_` function, this causes errors to be thrown later.

However, we have no way currently to wait for the promise within the `submit_` function. And `yield macroTask()`, requires a generator function that doesn't neccesarily play nice with `expect()`.

Was wondering if you would be open to having `handleSubmitEvent_` return a promise, that could be returned from the `Promise.all` in the `submit_` event? Or we could follow the pattern from `this.xhrSubmitPromise_`.

My main concern, I don't want to be shipping down code that will mostly / only be used testing by @zhouyx suggestion. Though It seems `this.xhrSubmitPromise_` is only used for this purpose.

Thanks!

**Comment by cvializ** [comment]
Sure, I think it should be fine to have `submit_` return its Promise. Adding code for testing shouldn't be shipped in production, since the `@visibleForTesting` annotation or `if (getMode().test) { /* ... */ }` should be removed by tree-shaking at compile time. It's always good to verify that that is working as expected though

File: extensions/amp-form/0.1/amp-form.js [hunk 16]
```diff
@@ -402,17 +410,15 @@ export class AmpForm {
    * Handles submissions through action service invocations.
    *   e.g. <img on=tap:form.submit>
    * @param {!../../../src/service/action-impl.ActionInvocation} invocation
+   * @return {?Promise}
    * @private
    */
   handleSubmitAction_(invocation) {
     if (this.state_ == FormState.SUBMITTING || !this.checkValidity_()) {
-      return;
+      return null;
```
Comment (cvializ) [inline extensions/amp-form/0.1/amp-form.js]: It's an antipattern for a function that returns a Promise to also return a synchronous value. It prevents chaining from working correctly, and mixes synchronous and asynchronous behavior ([which releases Zalgo](http://blog.izs.me/post/59142742143/designing-apis-for-asynchrony)).

Bad:
```js
/**
 * The major drawbacks here are:
 * 1. You can forget to check if the result is null and have a null reference error
 * 2. The function sometimes returns synchronously and sometimes asynchronously, 
 *      so timing expectations are weird
 * @param {boolean} flag
 * @return {?Promise}
 */
function doStuff(flag) { 
  if (flag) {
    return Promise.resolve(successValue);
  } else {
    return null;
  }
}

const result = doStuff(false);
if (result) {
  doStuff().then(() => doMoreStuff());
}
```

Good:
```js
function doStuff(flag) {
  if (flag) {
    Promise.resolve(successValue);
  } else {
    Promise.resolve(null);
  }
}

const result = doStuff(false).then(result => {
  if (result) { 
    return doMoreStuff();
  }
});
```

Comment (torch2424) [reply extensions/amp-form/0.1/amp-form.js]: I've learned more from this PR than my 5 years in college during my CS degree 😂 

Thank you! Will make the changes 😄

**Comment by torch2424** [comment]
@cvializ Made requested changes, and finished tests 😄 

Let me know if you see any other tests I should write. And I will continue debugging from the email thread we started about flaky tests on my machine.

Thank you!

P.S My apologies for the PR growing to be so big 😢

File: extensions/amp-form/0.1/amp-form.js [hunk 17]
```diff
@@ -488,14 +499,48 @@ export class AmpForm {
...
+    // Promises to run before submitting the form
+    const presubmitPromises = [];
+    presubmitPromises.push(this.doVarSubs_(varSubsFields));
+    iterateCursor(asyncInputs, asyncInput => {
+      presubmitPromises.push(this.getValueForAsyncInput_(asyncInput));
+    });
+
+    return this.waitOnPromisesOrTimeout_(presubmitPromises, 10000).then(
+        () => this.handlePresubmitSuccess_(trust),
+        error => this.handlePresubmitError_(/**@type {!Error}*/(error))
```
Comment (zhouyx) [inline extensions/amp-form/0.1/amp-form.js]: double confirming. I see `this.handlePresubmitError_` also creates a promise. So we don't need to wait for that promise here right?

File: extensions/amp-form/0.1/amp-form.js [hunk 18]
```diff
@@ -488,14 +499,48 @@ export class AmpForm {
...
+    this.setState_(FormState.SUBMITTING);
+
+    // Promises to run before submitting the form
+    const presubmitPromises = [];
+    presubmitPromises.push(this.doVarSubs_(varSubsFields));
+    iterateCursor(asyncInputs, asyncInput => {
+      presubmitPromises.push(this.getValueForAsyncInput_(asyncInput));
+    });
+
+    return this.waitOnPromisesOrTimeout_(presubmitPromises, 10000).then(
```
Comment (zhouyx) [inline extensions/amp-form/0.1/amp-form.js]: nit: Could we make the 10000 a const? It's easier to understand, also to change value later.

File: extensions/amp-recaptcha-input/0.1/test/test-amp-recaptcha-input.js [hunk 19]
```diff
@@ -32,19 +32,26 @@ describes.realWin('amp-recaptcha-input', {
     toggleExperiment(win, 'amp-recaptcha-input', true);
   });
 
-  function getRecaptchaInput(skipSiteKey, skipAction) {
+  function getRecaptchaInput(config = {}) {
```
Comment (zhouyx) [inline extensions/amp-recaptcha-input/0.1/test/test-amp-recaptcha-input.js]: I find `createElementWithAttributes` could be helpful here. We could make the input the attributes object. 
The `skipName` `skipSitekey`.. can be a bit confusing.

File: extensions/amp-form/0.1/amp-form.js (hunk 17 above)
Comment (torch2424) [reply extensions/amp-form/0.1/amp-form.js]: Correct. `this.submit_` used to have pieces of async within it, though it was called sync. And this error handling had promises that weren't returned as well.

**Comment by torch2424** [comment]
@zhouyx @cvializ 

Fixed all amp-form tests, added async input tests, and did a huge refactor to remove duplicated, or badly formatted tests to all handle asynchrony.

Let me know what you think! Also, super apologies on the gigantic PR. Most of it is tests at this point, and the Implementation was already approved, so I think just taking a look at that should be fine 😢 👍

File: extensions/amp-form/0.1/amp-form.js [hunk 20]
```diff
@@ -652,6 +736,33 @@ export class AmpForm {
...
+  getValueForAsyncInput_(asyncInput) {
+    return asyncInput.getImpl()
+        .then(implementation => implementation.getValue())
+        .then(value => {
+          const name = asyncInput.getAttribute(AsyncInputAttributes.NAME);
+          let input = this.form_.querySelector(
+              `input[name=${escapeCssSelectorIdent(name)}]`
+          );
+          if (!input) {
+            input = document.createElement('input');
```
Comment (zhouyx) [inline extensions/amp-form/0.1/amp-form.js]: use `this.win_.document` for easier testing, and prevent cases where `this.win_ != window`. 
And please use `#createElementsWithAttributes`.

File: extensions/amp-recaptcha-input/0.1/test/test-amp-recaptcha-input.js [hunk 21]
```diff
@@ -32,19 +33,24 @@ describes.realWin('amp-recaptcha-input', {
...
-          'fake-sitekey-fortesting');
-    }
-    if (!skipAction) {
-      ampRecaptchaInput.setAttribute('data-action',
-          'unit-testing');
-    }
+  const ampRecaptchaInputAttributes = {
+    layout: 'nodisplay',
+    name: 'recaptcha-test',
+    'data-sitekey': 'fake-sitekey-fortesting',
```
Comment (zhouyx) [inline extensions/amp-recaptcha-input/0.1/test/test-amp-recaptcha-input.js]: nit: since you're using string key, can we convert to 'layout' and 'name', just to match style : )

File: src/async-input.js [hunk 22]
```diff
@@ -39,14 +39,35 @@ export class AsyncInput {
...
+ */
+export const AsyncInputAttributes = {
+  /**
+   * data-name
+   *
+   * Required attribute that must be asserted by every async-input
+   * Element. This is used by AMP form to add the key
+   * for the form submission request
+  */
+  'NAME': 'name',
```
Comment (zhouyx) [inline src/async-input.js]: nit: NAME since you're using `AsyncInputAttributes.NAME` elsewhere, so that closure compiler can optimize the key string.

File: extensions/amp-form/0.1/amp-form.js [hunk 23]
```diff
@@ -488,14 +505,51 @@ export class AmpForm {
...
+    // Promises to run before submitting the form
+    const presubmitPromises = [];
+    presubmitPromises.push(this.doVarSubs_(varSubsFields));
+    iterateCursor(asyncInputs, asyncInput => {
+      presubmitPromises.push(this.getValueForAsyncInput_(asyncInput));
+    });
+
+    return this.waitOnPromisesOrTimeout_(
+        presubmitPromises,
+        SUBMIT_TIMEOUT
```
Comment (zhouyx) [inline extensions/amp-form/0.1/amp-form.js]: I noticed one thing here. The wait for var substitution is 100ms, while the wait for async input is 10000ms. However we are combining the wait for varSubsFields and asyncInput here, and the timeout is 10000ms for all. Which could lead to a 10s wait for var substitution. 

 to @cvializ to decide whether this is ok or not.

**Comment by torch2424** [comment]
Since these are to @cvializ , should I just fix these changes, or wait for their input? @zhouyx 😮

**Comment by zhouyx** [comment]
I'll wait for @cvializ's review. Since most changes lives within amp-form. Thanks.

File: extensions/amp-form/0.1/amp-form.js (hunk 23 above)
Comment (cvializ) [reply extensions/amp-form/0.1/amp-form.js]: The variable substitution will still timeout at 100ms since it also calls `waitOnPromisesOrTimeout`. 10000ms should be fine to begin with, and we can shorten it later if it ends up being too long.

File: extensions/amp-form/0.1/amp-form.js (hunk 20 above)
Comment (cvializ) [reply extensions/amp-form/0.1/amp-form.js]: +1, never use raw `document` or `window` references in AMP components. We should really make this a linter rule.

File: extensions/amp-form/0.1/amp-form.js [hunk 24]
```diff
@@ -488,14 +505,51 @@ export class AmpForm {
...
+    // Do any assertions we may need to do
+    // For NonXhrGET
+    // That way we can determine if
+    // we can submit synchronously
+    if (!this.xhrAction_ && this.method_ == 'GET') {
+      this.assertSsrTemplate_(false, 'Non-XHR GETs not supported.');
+      this.assertNoSensitiveFields_();
+
+      // If we have no async inputs, we can just submit synchronously
+      if (asyncInputs.length === 0) {
```
[the remark this answers is not in the document]
Comment (cvializ) [reply extensions/amp-form/0.1/amp-form.js]: More branching definitely introduces more complexity, but I think that amp-form will need a refactor soon, and we can address this difficulty in a future PR.

**Comment by cvializ** [comment]
Left a couple comments. It's getting close to ready for review but these issues do need to be addressed before LGTM

**Comment by torch2424** [comment]
@zhouyx @cvializ

Fixed all the comments (I think?) that required changes.

Let me know what you think / how to move forward 😄

File: extensions/amp-form/0.1/amp-form.js [hunk 25]
```diff
@@ -508,6 +563,39 @@ export class AmpForm {
...
+   * @param {!ActionTrust} trust
+   * @return {!Promise}
+   */
+  handlePresubmitSuccess_(trust) {
+    if (this.xhrAction_) {
+      return this.handleXhrSubmit_(trust);
+    } else if (this.method_ == 'POST') {
+      this.handleNonXhrPost_();
+    } else if (this.method_ == 'GET') {
+      this.handleNonXhrGet_(event);
```
Comment (cvializ) [inline extensions/amp-form/0.1/amp-form.js]: Is this the `window.event` variable? I don't see it defined. Using the global `event` should be avoided

https://developer.mozilla.org/en-US/docs/Web/API/Window/event

Comment (torch2424) [reply extensions/amp-form/0.1/amp-form.js]: Ah! You are complete correct! Must have forgot to pass that down when I moved this into its own function 😮 Thanks! Will fix that now! Surprised check types didn't catch this

**Comment by torch2424** [comment]
@cvializ This is good to go! Fixed all comments 😄

**Comment by torch2424** [comment]
@cvializ 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 

Gonna fix the linting errors, and then merge this in! 😄
